storage/engine/mvcc.go

```diff
@@ -653,54 +653,47 @@ func mvccPutInternal(engine Engine, ms *proto.MVCCStats, key proto.Key, timestam
 		return err
 	}
 
-	var newMeta *proto.MVCCMetadata
-	// In case the key metadata exists.
 	if ok {
-		// There is an uncommitted write intent and the current Put
-		// operation does not come from the same transaction.
-		// This should not happen since range should check the existing
-		// write intent before executing any Put action at MVCC level.
-		if meta.Txn != nil && (txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID)) {
-			return &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
-		}
+		// There is existing metadata for this key; ensure our write is permitted.
 
-		// We can update the current metadata only if both the timestamp
-		// and epoch of the new intent are greater than or equal to
-		// existing. If either of these conditions doesn't hold, it's
-		// likely the case that an older RPC is arriving out of order.
-		//
-		// Note that if meta.Txn!=nil and txn==nil, a WriteIntentError was
-		// returned above.
-		if !timestamp.Less(meta.Timestamp) &&
-			(meta.Txn == nil || txn.Epoch >= meta.Txn.Epoch) {
-			// If this is an intent and timestamps have changed,
-			// need to remove old version.
-			if meta.Txn != nil && !timestamp.Equal(meta.Timestamp) {
+		if meta.Txn != nil {
+			// There is an uncommitted write intent.
+
+			if txn == nil || !bytes.Equal(meta.Txn.ID, txn.ID) {
+				// The current Put operation does not come from the same
+				// transaction.  This should not happen since range should check
+				// the existing write intent before executing any Put action at
+				// MVCC level.
+				return &proto.WriteIntentError{Intents: []proto.WriteIntentError_Intent{{Key: key, Txn: *meta.Txn}}}
+			} else if txn.Epoch < meta.Txn.Epoch || timestamp.Less(meta.Timestamp) {
+				// Ignore old writes from the current transaction; it's just an
+				// RPC arriving out of order.
+				return nil
+			} else if !timestamp.Equal(meta.Timestamp) {
+				// We are replacing our own write intent.
 				versionKey := mvccEncodeTimestamp(metaKey, meta.Timestamp)
 				if err := engine.Clear(versionKey); err != nil {
 					return err
 				}
```

Comment:
Yes: if none of these if statements are true, then we are replacing an intent from our same transaction which was written with the same timestamp (so we don't need an explicit clear of the old value; we'll just overwrite it directly). However, this is concerning in combination with the out-of-order comments above: if a client does send two different writes to the same key with the same timestamp, we have no way of knowing which one was intended to be later if commands may get out of order. But I'm not sure if that's actually possible; I think there is higher-level logic that will prevent out-of-order RPCs from reaching this point (unless perhaps they're sent as a part of the same batch). @spencerkimball, what is the concern about reordered RPCs at this level?